File: iosgl/iosgl/bt/display/line/LineDisplaySprite.m
//
//  LineDisplaySprite.m
//  iosgl
//
//  Created by zhao on 6/3/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Context3D.h"
#import "Scene3D.h"
#import "LineDisplayShader.h"
#import "LineDisplaySprite.h"
#import "ProgrmaManager.h"

@interface LineDisplaySprite ()

@property (nonatomic, strong) NSMutableArray<Vector3D *>*  linePointArr;
 


@end
@implementation LineDisplaySprite

- (instancetype)init
{
    self = [super init];
    if (self) {
         self.colorV3d=[[Vector3D alloc]x:1 y:0 z:0];
    }
    return self;
}
-(void)clearLine;
{
    if(self.linePointArr&&self.linePointArr.count){
        [self.linePointArr removeAllObjects];
    }
}
-(void)addLineA2B:(Vector3D*)a b:(Vector3D*)b color:(Vector3D*)color;
{
    self.colorV3d=color;
    [self addLineA2B:a b:b];
}
-(void)addLineA2B:(Vector3D*)a b:(Vector3D*)b;
{
    if(!self.linePointArr ){
        self.linePointArr  =[[NSMutableArray alloc]init];
    }
    [self.linePointArr addObject:a];
    [self.linePointArr addObject:[[Vector3D alloc]x:self.colorV3d.x y:self.colorV3d.y z:self.colorV3d.z]];
    [self.linePointArr addObject:b];
    [self.linePointArr addObject:[[Vector3D alloc]x:self.colorV3d.x y:self.colorV3d.y z:self.colorV3d.z]];
}
-(void)onCreated;
{
    self.objData=[[ObjData alloc]init];
    [self clearLine];
    
    self.colorV3d=[[Vector3D alloc]x:1 y:0 z:0];
    [self addLineA2B:[[Vector3D alloc]x:0 y:0 z:0]   b:[[Vector3D alloc]x:100.0f y:0 z:0]];
    self.colorV3d=[[Vector3D alloc]x:0 y:0 z:1];
    [self addLineA2B:[[Vector3D alloc]x:0 y:0 z:0]   b:[[Vector3D alloc]x:0 y:0 z:100.0f]];
}
-(void)registetProgame;
{
     [[ProgrmaManager default] registe:LineDisplayShader.shaderStr shader3d: [[LineDisplayShader alloc]init]];
      self.shader3d=  [[ProgrmaManager default] getProgram:LineDisplayShader.shaderStr];
}
-(void)refrishLineDataToGpu;
{
    if(self.linePointArr&&self.linePointArr.count){
        GLfloat attrArr[self.linePointArr.count*3];
        for (int i=0; i<self.linePointArr.count; i++) {
            attrArr[i*3+0]= self.linePointArr[i].x;
            attrArr[i*3+1]= self.linePointArr[i].y;
            attrArr[i*3+2]= self.linePointArr[i].z;
        }
        unsigned int Indices[self.linePointArr.count];
        for (int j=0; j<self.linePointArr.count/2; j++) {
             Indices[j]=j;
        }
        GLuint verticesBuffer;
        glGenBuffers(1, &verticesBuffer);
        glBindBuffer(GL_ARRAY_BUFFER, verticesBuffer);
        glBufferData(GL_ARRAY_BUFFER, sizeof(attrArr), attrArr, GL_DYNAMIC_DRAW);
        self.objData.verticesBuffer=verticesBuffer;
        GLuint indexBuffer;
        glGenBuffers(1, &indexBuffer);
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, indexBuffer);
        glBufferData(GL_ELEMENT_ARRAY_BUFFER, sizeof(Indices), Indices, GL_STATIC_DRAW);
        self.objData.indexBuffer=indexBuffer;
        self.objData.trinum=(int)self.linePointArr.count/2;
        [self clearLine];
    }
}

-(void)upFrame{
    
    if(self.shader3d&&self.objData){
        [self refrishLineDataToGpu];
        GLuint progame= self.shader3d.program;
        glUseProgram(progame);
    
        [self setVc];
        [self setVa];
        
        glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, self.objData.indexBuffer);
        glDrawElements(GL_LINES, self.objData.trinum, GL_UNSIGNED_INT, 0);
        
    }
 
}
-(void)setVc;
{
    Context3D *context3D=self.scene3d.context3D;
    [context3D setVcMatrix4fv:self.shader3d name:"viewMatrix" data:self.viewMatrix.m];
    [context3D setVcMatrix4fv:self.shader3d name:"posMatrix" data:self.posMatrix3d.m];
}
-(void)setVa;
{
    Context3D *context3D=self.scene3d.context3D;
    [context3D pushVa:self.objData.verticesBuffer];
    [context3D setVaOffset:self.shader3d name:"position" dataWidth:3 stride:24 offset:0];
    [context3D setVaOffset:self.shader3d name:"attcolor" dataWidth:3 stride:24 offset:3];
    
}



@end

File: androld/app/src/main/java/z3d/display/line/LineDisplaySprite.java
package z3d.display.line;

import android.opengl.GLES20;

import java.util.ArrayList;

import z3d.base.ObjData;
import z3d.core.Context3D;
import z3d.display.basedis.DisplayBaseShader;
import z3d.display.basedis.DisplayBaseSprite;
import z3d.program.ProgrmaManager;
import z3d.vo.Matrix3D;
import z3d.vo.Vector3D;

public class LineDisplaySprite extends DisplayBaseSprite {


//    @property (nonatomic, strong) Vector3D* colorV3d;
//-(void)clearLine;
//-(void)addLineA2B:(Vector3D*)a b:(Vector3D*)b;

    private Vector3D colorV3d;


    protected void  registetProgame()
    {

        ProgrmaManager.getInstance().registe(LineDisplayShader.shaderNameStr,new LineDisplayShader());
        this.shader3D=ProgrmaManager.getInstance().getProgram(LineDisplayShader.shaderNameStr);

    }
    public  LineDisplaySprite()
    {
        super();
        this.onCreated();
    }
    protected void  onCreated()
    {

        this.clearLine();
        this.colorV3d=new Vector3D(1,0,0);

        this.addLineA2B(new Vector3D(0,0 ,0) ,new Vector3D(100,0,0));
        this.addLineA2B(new Vector3D(0,0 ,0) ,new Vector3D(0,100,0));

    }

    public  void clearLine()
    {

    }
    public  void  addLineA2B(Vector3D a, Vector3D b)
    {

    }
    protected void  makeTempObjData()
    {
        this.objData =new ObjData();

        ObjData od=this.objData;

        od.verticeslist=new ArrayList<Float>();//结果顶点坐标列表
        od.verticeslist.add(0f);
        od.verticeslist.add(0f);
        od.verticeslist.add(0f);

        od.verticeslist.add(100f);
        od.verticeslist.add(0f);
        od.verticeslist.add(0f);




        od.indexs=new ArrayList<Short>();
        od.indexs.add((short)0);
        od.indexs.add((short)1);




        od.upToGup();


    }

    public void upFrame(){
        Context3D ctx=this.scene3d.context3D;

        if(this.shader3D!=null){

            this.modeMatrix.appendRotation(1, Vector3D.Z_AXIS);
            ctx.setProgame(this.shader3D.program);

            Matrix3D m=new Matrix3D();
            m.appendScale(10,10,0);


            ctx.setVcMatrix4fv(this.shader3D,"vpMatrix3D",this.scene3d.camera3D.modelMatrix.m);
            ctx.setVcMatrix4fv(this.shader3D,"posMatrix",this.modeMatrix.m);
            ctx.setVaOffset(this.shader3D, "vPosition");
            ctx.setVa(0,3,this.objData.vertexBuffer);



         //  GLES20.glDrawElements(GLES20.GL_TRIANGLES,this.objData.treNum, GLES20.GL_UNSIGNED_SHORT,this.objData.indexBuffer);

           GLES20.glDrawElements(GLES20.GL_LINES,2, GLES20.GL_UNSIGNED_SHORT,this.objData.indexBuffer);



            GLES20.glDisableVertexAttribArray(0);
        }


    }
}
